### PR TITLE
Add general and Java specific style guidelines

### PR DESCRIPTION
**Pull request by petroav**
This PR is the result of a style discussion we had while @martint and @dain were visiting the Teradata Boston office. I hope I didn't miss anything and if I did let me know and I'd be happy to do the appropriate edits.

I will make another PR to the Presto repo for the more Presto specific style guidelines.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,6 +8,38 @@ To install, copy `IntelliJIdea14/Airlift.xml` into `~/Library/Preferences/Intell
...
+
+## General Java style guidelines
+
+* Do not abbreviate Java variable names: `command` and not `cmd`.
+* Google's Guava is seen as an extension of the JDK. You should use it when
+    * certain functionality doesn't exist in the JDK or
+    * Guava presents you with a clearly superior API and doesn't degrade performance.
+
+  For example, do not use `Guava.checkNotNull()` because the same functionality exists
+  in the JDK as `Objects.requireNonNull()`.
```
Comment (hgschmie) [inline README.md]: it is ironic that this is the most quoted example while JDK requireNonNull is not equivalent to checkNotNull (it does not support the parameter templating). Using `requireNonNull(String.format(...))` next to the other Guava checks (`checkState` and `checkArgument`) makes formatting of error messages uneven and counterintuitive. In fact, the Guava developers state (http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/base/Preconditions.html):

--- cut ---
java.util.Objects.requireNonNull()

Projects which use com.google.common should generally avoid the use of Objects.requireNonNull(Object). Instead, use whichever of checkNotNull(Object) or Verify.verifyNotNull(Object) is appropriate to the situation. (The same goes for the message-accepting overloads.)
--- cut ---

so this recommendation also goes against the recommendations in the "Guava explained" documentation: https://code.google.com/p/guava-libraries/wiki/PreconditionsExplained 

I would suggest to be pragmatic here, nof dogmatic and follow the Guava documentation (with this one cited example being the obvious exception).

And yes, the modernizer plugin got that one wrong, too.

File: README.md [hunk 2]
```diff
@@ -8,6 +8,38 @@ To install, copy `IntelliJIdea14/Airlift.xml` into `~/Library/Preferences/Intell
 
 To install, copy `IntelliJIdea14/Airlift.xml` into `$HOME/.IdeaIC14/config/codestyles`
 
+## Programming guidelines
+
+* Commit messages should follow the format described [here](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html). Ideally, you should be able to look at the commit and:
```
Comment (electrum) [inline README.md]: Use this link instead: http://chris.beams.io/posts/git-commit/

File: README.md (hunk 1 above)
Comment (petroav) [inline README.md]: I have no strong opinion on this and your argument is completely valid. However, if I recall that is the example either @dain or @martint gave me.

What specific changes would you like to see? I can leave the two bullet points about Guava as they are and just remove the example about `Objects.requireNonNull()`. Is that OK?

File: README.md (hunk 2 above)
Comment (petroav) [inline README.md]: Done.

**Comment by petroav** [comment]
Updated PR with new commit addressing @electrum's comment and other minor cosmetic changes.

**Comment by petroav** [comment]
@electrum and @hgschmie can you take another look at this when you get a chance?

**Comment by findepi** [comment]
Can we consider this superseded by @kokosing's https://github.com/prestosql/presto/pull/4699?

**Pull request closed without merging**